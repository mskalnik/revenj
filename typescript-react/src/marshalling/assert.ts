
import { TypescriptResultSet } from '../ResultSet/ResultSet';
import {
  INT_MAX_VALUE,
  INT_MIN_VALUE,
  SHORT_MAX_VALUE,
  SHORT_MIN_VALUE,
  LONG_MAX_VALUE,
  LONG_MIN_VALUE,
} from '../constants';
import { isObject } from '../util/FunctionalUtils/FunctionalUtils';
import { formatNumber } from '../util/Formatters/NumberFormatter';
import { isNumber } from '../util/NumberUtils/NumberUtils';

type Dictionary<K extends string | number, V> = K extends string ? { [key: string]: V }: { [key: number]: V };

export const assertionWithErrorPrefix = <T>(prefix: string) => (assertion: (value: T) => T): (value: T) => T =>
  (value: T): T => {
    try {
      return assertion(value);
    } catch (error) {
      throw new Error(`${prefix}: ${error instanceof Error ? error.message : String(error)}`);
    }
  };

export const assertPresence = <T>(value: T, isNonNullable: boolean): T | undefined => {
  if (value == null) {
    if (isNonNullable) {
      throw new Error(`Value is required, but is not set!`);
    } else {
      return;
    }
  }

  return value;
};

export const assertPresenceStrict = <T>(value: T) => assertPresence(value, true);

export const assertString = <T extends string>(value: T, length?: number): T => {
  if (typeof value !== 'string') {
    throw new Error(`Expected a String, but value is ${value}`);
  }

  if (length != null && value.length > length!) {
    throw new Error(`Maximum length is ${length!}, but received string of length ${value.length}`);
  }

  return value;
};

export const assertInt = <T extends Int>(value: T): T => {
  const number = Number.parseInt(String(value), 10) as T;

  if (Number.isNaN(number) || value > INT_MAX_VALUE || value < INT_MIN_VALUE || typeof value !== 'number' || value !== number) {
    throw new Error(`Expected an Int, but value is ${value}`);
  }
  return number;
};

/**
 * Asserts that a given value is within the range of a 64-bit signed Long.
 * This is necessary because JavaScript loses numeric precision beyond Number.MAX_SAFE_INTEGER.
 * Ideally, serialization should be done using strings and deserialized into BigInt or a bignum,
 * but for now, this approach is acceptable.
 */
export const assertLong = <T extends Long>(value: T): T => {
  try {
    const number = BigInt(value);
    if (number < BigInt(LONG_MIN_VALUE) || number > BigInt(LONG_MAX_VALUE)) {
      throw new Error(`Expected a Long, but value is ${value}`);
    }
    if (number >= Number.MIN_SAFE_INTEGER && number <= Number.MAX_SAFE_INTEGER) {
      return Number.parseInt(String(value), 10) as unknown as T;
    }
    return number.toString() as unknown as T;
  } catch (error) {
    throw new Error(`Expected a Long, but value is ${value}`);
  }
};

export const assertShort = <T extends Short>(value: T): T => {
  const number = assertInt(value as unknown as Int);
  if (number > SHORT_MAX_VALUE || number < SHORT_MIN_VALUE) {
    throw new Error(`Number ${number} out of Short range`);
  }

  return number as unknown as T;
};

export const assertFloatingPoint = <T extends number>(value: T): T => {
  const number = Number.parseFloat(String(value)) as T;
  if (Number.isNaN(number) || typeof value !== 'number') {
    throw new Error(`Expected a floating-point number, but value is ${value}`);
  }
  return number;
};

export const assertDecimal = <T extends DecimalStr>(value: T): T => {
  const number = formatNumber(value) as T;
  if (!isNumber(number)) {
    throw new Error(`Expected a floating-point number, but value is ${value}`);
  }
  return number;
};

export const serializeBinary = (value: BinaryStr): BinaryStr =>
  btoa(
    encodeURIComponent(assertString(value))
    .replace(/%([0-9A-F]{2})/g, (_, g) => String.fromCharCode(parseInt(g, 16))),
  ) as BinaryStr;

export const deserializeBinary = (value: BinaryStr): BinaryStr =>
  decodeURIComponent(
    Array.from(atob(assertString(value)))
    .map((c) => '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2)).join(''),
  ) as BinaryStr;

export const assertArray = <T>(it: T[]): T[] => {
  if (!Array.isArray(it)) {
    throw new Error(`Expected an array, but value is ${it}`);
  }

  return Array.from(it);
};

export const serializeSet = <T>(it: Set<T>): T[] => {
  if (!(it instanceof Set)) {
    throw new Error(`Expected a Set, but value is ${it}`);
  }

  return Array.from(it);
};

export const deserializeSet = <T>(it: T[]): Set<T> => {
  if (!Array.isArray(it)) {
    throw new Error(`Expected an array, but value is ${it}`);
  }

  return new Set(it);
};

export const serializeMap = <K extends string | number, V>(it: Map<K, V>): Dictionary<K, V> => {
  if (it instanceof Map) {
    const ret: Dictionary<K, V> = {} as Dictionary<K, V>;
    it.forEach((value, key) => {
      ret[key as any] = value;
    });
    return ret;
  }

  if (isObject(it)) {
    return it;
  }

  throw new Error(`Expected a Map, but value is ${it}`);
};

export const deserializeMap = <K extends string | number, V>(it: Dictionary<K, V>): Map<K, V> => {
  // A... non-exhaustive list
  if (!isObject(it) || Array.isArray(it) || it instanceof Map || it instanceof Set) {
    throw new Error(`Expected a dictionary object, but value is ${it}`);
  }

  const ret = new Map<K, V>();
  Object.keys(it).forEach((key) => {
    ret.set(key as unknown as K, it[key]);
  });
  return ret;
};

export const assertIterable = <T>(it: Iterable<T>): T[] => {
  if (it == null || (typeof it[Symbol.iterator] !== 'function')) {
    throw new Error(`Expected an iterable collection, but value is ${it}`);
  }

  return Array.from(it);
};

export const serializeResultSet = (it: TypescriptResultSet): ResultSet => {
  return assertArray(TypescriptResultSet.serialize(it));
};

export const deserializeResultSet = (it: ResultSet): TypescriptResultSet => {
  return TypescriptResultSet.deserialize(assertArray(it));
};

export const assertEnum = <T>(values: T[], name: string) => (it: T, isNonNullable: boolean, _: string): T => {
  if (assertPresence(it, isNonNullable) == null && !isNonNullable) {
    return it;
  }

  if (!values.includes(it)) {
    throw new Error(`Value ${it} is not a valid member of enumeration ${name}`);
  }

  return it;
};

interface IDictionary {
  [key: string]: string;
}

export const assertDictionary = (value: any, field: string): IDictionary => {
  if (
    typeof value !== 'object' &&
    Object.keys(value).some((key) => typeof key !== 'string') &&
    Object.values(value).some((value) => typeof value !== 'string')
  ) {
    throw new Error(`${field} must be a Dictionary, received ${JSON.stringify(value)}`);
  }
  return value as IDictionary;
};

// Composite Type Assertions

export const assertS3 = (value: any): S3 => {
  assertionWithErrorPrefix<any>('S3 key')(assertString)(value.key);
  assertionWithErrorPrefix<any>('S3 bucket')(assertString)(value.bucket);
  assertionWithErrorPrefix<any>('S3 length')(assertLong)(value.length);

  return {
    bucket: assertionWithErrorPrefix<string>('S3 bucket')(assertString)(value.bucket),
    key: assertionWithErrorPrefix<string>('S3 key')(assertString)(value.key),
    length: assertionWithErrorPrefix<Long>('S3 length')(assertLong)(value.length),
    metadata: value.metadata != null ? assertDictionary(value.metadata, 'S3.metadata') : undefined,
    mimeType: value.mimeType != null ? assertionWithErrorPrefix<string>('S3 mime type')(assertString)(value.mimeType) : undefined,
    name: value.name != null ? assertionWithErrorPrefix<string>('S3 name')(assertString)(value.name) : undefined,
  };
};
