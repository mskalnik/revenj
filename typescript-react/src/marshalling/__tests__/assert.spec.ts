import {
  INT_MAX_VALUE,
  INT_MIN_VALUE,
  LONG_MAX_VALUE,
  LONG_MIN_VALUE,
} from '../../constants';
import {
  ColumnType,
  TypescriptResultSet,
} from '../../ResultSet/ResultSet';
import {
  SHORT_MAX_VALUE,
  SHORT_MIN_VALUE,
} from '../../constants';
import * as Assert from '../assert';
import { parseBigNum } from '../../util/NumberUtils/NumberUtils';

describe('data marshalling asserting', () => {
  describe('assertPresence', () => {
    it('should return a value if it is set', () => {
      expect(Assert.assertPresence(1, true)).toEqual(1);
      expect(Assert.assertPresence(1, false)).toEqual(1);
      expect(Assert.assertPresence([1], true)).toEqual([1]);
      expect(Assert.assertPresence([1], false)).toEqual([1]);
      expect(Assert.assertPresence('test', true)).toEqual('test');
      expect(Assert.assertPresence('test', false)).toEqual('test');
    });

    it('should return undefined if the value is not present, but not required', () => {
      expect(Assert.assertPresence(null, false)).not.toBeNull();
      expect(Assert.assertPresence(null, false)).toBeUndefined();
      expect(Assert.assertPresence(undefined as any, false)).toBeUndefined();
    });

    it('should raise an error if the value is not present, but required', () => {
      expect(() => Assert.assertPresence(null, true)).toThrow();
      expect(() => Assert.assertPresence(null, true)).toThrowErrorMatchingSnapshot();
      expect(() => Assert.assertPresence(undefined as any, true)).toThrow();
      expect(() => Assert.assertPresence(undefined as any, true)).toThrowErrorMatchingSnapshot();
    });
  });

  describe('assertString', () => {
    it('should pass through strings', () => {
      const valids = ['bla', '', `test
ing`];
      for (const valid of valids) {
        expect(() => Assert.assertString(valid)).not.toThrow();
        expect(Assert.assertString(valid)).toEqual(valid);

      }
    });

    it('should explode on non-strings', () => {
      const invalids = [null, 12, {}];
      for (const invalid of invalids) {
        expect(() => Assert.assertString(invalid as any)).toThrowErrorMatchingSnapshot();
      }
    });

    it('should also validate length if specified', () => {
      const valids = ['a', 'aa', 'aaa'];
      const invalids = ['aaaaaaa', '    a    '];
      for (const valid of valids) {
        expect(Assert.assertString(valid, 3)).toBe(valid);
      }

      for (const invalid of invalids) {
        expect(() => Assert.assertString(invalid, 3)).toThrowErrorMatchingSnapshot();
      }
    });
  });

  describe('assertInt', () => {
    it('should pass through valid integers', () => {
      const valids = [INT_MIN_VALUE, -10, 0, 235, INT_MAX_VALUE] as Int[];
      for (const valid of valids) {
        expect(Assert.assertInt(valid)).toBe(valid);
      }
    });

    it('should explode on non-integers', () => {
      const invalids = [INT_MIN_VALUE - 1, 'bla', null, 1.55, INT_MAX_VALUE + 1] as any[];
      for (const invalid of invalids) {
        expect(() => Assert.assertInt(invalid)).toThrowErrorMatchingSnapshot();
      }
    });

    it.skip('should assert valid Int range', () => {}); // TODO: Some nice day
  });

  describe('assertLong', () => {
    it('should pass through valid longs', () => {
      const valids = [
        '-10',
        '0',
        '235',
        '-1',
        '1',
        '-9007199254740995',
        '-9007199254740999',
        '9007199254740995',
        '9007199254740999',
        Number.MIN_SAFE_INTEGER.toString(),
        Number.MAX_SAFE_INTEGER.toString(),
        LONG_MAX_VALUE.toString(),
        LONG_MIN_VALUE.toString(),
      ] as Long[];
      for (const valid of valids) {
        expect(Assert.assertLong(valid)).toBe(valid);
      }
    });

    it('should explode on non-longs', () => {
      const invalids = [
        'bla',
        null,
        1.55,
        Number.MAX_VALUE,
        Number.MIN_VALUE,
        Number.POSITIVE_INFINITY,
        Number.NEGATIVE_INFINITY,
      ] as any[];
      for (const invalid of invalids) {
        expect(() => Assert.assertLong(invalid)).toThrowErrorMatchingSnapshot();
      }
    });

    it('should assert valid Long range', () => {
      const minValid = LONG_MIN_VALUE.toString();
      const maxValud = LONG_MAX_VALUE.toString();
      expect(Assert.assertLong(minValid)).toBe(minValid);
      expect(Assert.assertLong(maxValud)).toBe(maxValud);

      const bellowMin = parseBigNum(LONG_MIN_VALUE).minus(1).toString();
      const aboveMax = parseBigNum(LONG_MAX_VALUE).plus(1).toString();
      expect(() => Assert.assertLong(bellowMin)).toThrowErrorMatchingSnapshot();
      expect(() => Assert.assertLong(aboveMax)).toThrowErrorMatchingSnapshot();
    });
  });

  describe('assertShort', () => {
    it('should pass through valid shorts', () => {
      const valids = [-10, 0, 235] as Short[];
      for (const valid of valids) {
        expect(Assert.assertShort(valid)).toBe(valid);
      }
    });

    it('should explode on non-shorts', () => {
      const invalids = ['bla', null, 1.55];
      for (const invalid of invalids) {
        expect(() => Assert.assertShort(invalid as any)).toThrowErrorMatchingSnapshot();
      }
    });

    it('should assert valid Short range', () => {
      const inRange = [SHORT_MIN_VALUE, SHORT_MAX_VALUE] as Short[];
      const outOfRange = [SHORT_MIN_VALUE - 1, SHORT_MAX_VALUE + 1];

      for (const ok of inRange) {
        expect(Assert.assertShort(ok)).toBe(ok);
      }

      for (const bad of outOfRange) {
        expect(() => Assert.assertShort(bad as any)).toThrowErrorMatchingSnapshot();
      }
    });
  });

  describe('serializeBinary', () => {
    it('should encode valid strings', () => {
      expect(Assert.serializeBinary('test' as BinaryStr)).toEqual('dGVzdA==');
    });

    it('should explode on non-strings', () => {
      const invalids = [null, 2, {}] as any[];

      for (const invalid of invalids) {
        expect(() => Assert.serializeBinary(invalid)).toThrowErrorMatchingSnapshot();
      }
    });
  });

  describe('deserializeBinary', () => {
    it('should decode valid strings', () => {
      expect(Assert.deserializeBinary('dGVzdA==' as BinaryStr)).toEqual('test');
    });

    it('should explode on non-strings', () => {
      const invalids = [null, 2, {}] as any[];

      for (const invalid of invalids) {
        expect(() => Assert.deserializeBinary(invalid)).toThrowErrorMatchingSnapshot();
      }
    });
  });

  describe('assertArray', () => {
    it('should pass through arrays', () => {
      const arrays = [
        [],
        [1, 2, 3],
        ['bla', 'bli', 'blu'], // Thanks, Pauline
      ];
      for (const array of arrays) {
        expect(() => Assert.assertArray<any>(array)).not.toThrow();
        expect(Assert.assertArray<any>(array)).toEqual(array);
      }
    });

    it('should explode on non-arrays', () => {
      const badApples = [
        null,
        {},
        new Set([]),
      ];

      for (const badApple of badApples) {
        expect(() => Assert.assertArray<any>(badApple as any)).toThrow();
        expect(() => Assert.assertArray<any>(badApple as any)).toThrowErrorMatchingSnapshot();
      }
    });

    it('should make a copy of the original value', () => {
      const original = [1, 2, 3];
      const copy = Assert.assertArray(original);
      expect(original).toEqual(copy);
      original.push(4);
      expect(original).not.toEqual(copy);
      expect(original).toHaveLength(4);
      expect(copy).toHaveLength(3);
    });
  });

  describe('serializeSet', () => {
    it('should serialize sets into arrays', () => {
      expect(Assert.serializeSet(new Set([]))).toEqual([]);
      expect(Assert.serializeSet(new Set([1, 2, 3]))).toEqual([1, 2, 3]);
      expect(Assert.serializeSet(new Set(['bla', 'bli', 'blu']))).toEqual(['bla', 'bli', 'blu']);
    });

    it('should explode on non-sets', () => {
      const badApples = [
        null,
        {},
        [],
      ];

      for (const badApple of badApples) {
        expect(() => Assert.serializeSet<any>(badApple as any)).toThrow();
        expect(() => Assert.serializeSet<any>(badApple as any)).toThrowErrorMatchingSnapshot();
      }
    });
  });

  describe('deserializeSet', () => {
    it('should deserialize arrays into sets', () => {
      expect(Assert.deserializeSet([])).toEqual(new Set([]));
      expect(Assert.deserializeSet([1, 2, 3])).toEqual(new Set([1, 2, 3]));
      expect(Assert.deserializeSet(['bla', 'bli', 'blu'])).toEqual(new Set(['bla', 'bli', 'blu']));
    });

    it('should explode on non-arrays', () => {
      const badApples = [
        null,
        {},
        new Set([]),
      ];

      for (const badApple of badApples) {
        expect(() => Assert.deserializeSet<any>(badApple as any)).toThrow();
        expect(() => Assert.deserializeSet<any>(badApple as any)).toThrowErrorMatchingSnapshot();
      }
    });
  });

  describe('serializeMap', () => {
    it('should serialize maps into lookups', () => {
      const strMap = new Map();
      strMap.set('a', 'test');
      strMap.set('b', 'test2');

      const intMap = new Map();
      intMap.set(1, 'test');
      intMap.set(2, 'test2');

      expect(Assert.serializeMap<any, any>(new Map())).toEqual({});
      expect(Assert.serializeMap<any, any>(strMap)).toEqual({
        a: 'test',
        b: 'test2',
      });
      expect(Assert.serializeMap<any, any>(intMap)).toEqual({
        [1]: 'test',
        [2]: 'test2',
      });
    });

    it('should explode on non-maps', () => {
      const badApples = [
        null,
        [],
      ];

      for (const badApple of badApples) {
        expect(() => Assert.serializeMap<any, any>(badApple as any)).toThrow();
        expect(() => Assert.serializeMap<any, any>(badApple as any)).toThrowErrorMatchingSnapshot();
      }
    });
  });

  describe('deserializeMap', () => {
    it('should deserialize lookups into maps', () => {
      const strMap = new Map();
      strMap.set('a', 'test');
      strMap.set('b', 'test2');

      expect(Assert.deserializeMap({})).toEqual(new Map());
      expect(Assert.deserializeMap({
        a: 'test',
        b: 'test2',
      })).toEqual(strMap);
    });

    it('should explode on non-lookups', () => {
      const badApples = [
        null,
        [],
        new Map(),
        new Set([]),
        () => { return; },
      ];

      for (const badApple of badApples) {
        expect(() => Assert.deserializeMap(badApple as any)).toThrow();
        expect(() => Assert.deserializeMap(badApple as any)).toThrowErrorMatchingSnapshot();
      }
    });
  });

  describe('serializeResultSet', () => {
    const rsOverTheWire = [['a', 'test'], ['String', 'String'], ['test', 'test'], ['test1', 'test1']];
    const rs = new TypescriptResultSet(['a', 'test'], [ColumnType.String, ColumnType.String], [['test', 'test'], ['test1', 'test1']]);

    it('should turn it into an array of objects', () => {
      expect(() => Assert.serializeResultSet(rs)).not.toThrow();
      expect(Assert.serializeResultSet(rs)).toEqual(rsOverTheWire);
    });

    it('should explode on invalid shapes', () => {
      const badApples: any[] = [
        {},
        [{}],
        [[]],
      ];

      for (const badApple of badApples) {
        expect(() => Assert.serializeResultSet(badApple)).toThrow();
        expect(() => Assert.serializeResultSet(badApple)).toThrowErrorMatchingSnapshot();
      }

    });
  });

  describe('deserializeResultSet', () => {
    const rsOverTheWire = [['a', 'test'], ['String', 'String'], ['test', 'test'], ['test1', 'test1']];
    const rs = new TypescriptResultSet(['a', 'test'], [ColumnType.String, ColumnType.String], [['test', 'test'], ['test1', 'test1']]);

    it('should turn it into an array of objects', () => {
      expect(() => Assert.deserializeResultSet(rsOverTheWire)).not.toThrow();
      expect(Assert.deserializeResultSet(rsOverTheWire)).toEqual(rs);
    });

    it('should explode on invalid shapes', () => {
      const badApples: any[] = [
        {},
        [new Map()],
        [[]],
      ];

      for (const badApple of badApples) {
        expect(() => Assert.deserializeResultSet(badApple)).toThrow();
        expect(() => Assert.deserializeResultSet(badApple)).toThrowErrorMatchingSnapshot();
      }

    });
  });

  describe('assertEnum', () => {
    it('should test that the value is one of the provided values', () => {
      const enumChecker = Assert.assertEnum(['Active', 'Inactive'], 'Status');

      expect(() => enumChecker('Active', true, '')).not.toThrow();
      expect(() => enumChecker('Inactive', true, '')).not.toThrow();
      expect(() => enumChecker(undefined as any, false, '')).not.toThrow();
      expect(() => enumChecker('Other', true, '')).toThrowErrorMatchingSnapshot();
      expect(() => enumChecker('INACTIVE', true, '')).toThrowErrorMatchingSnapshot();
      expect(() => enumChecker(undefined as any, true, '')).toThrowErrorMatchingSnapshot();
    });

    it('should provide the passed in type name on failure', () => {
      const enumChecker = Assert.assertEnum(['Active', 'Inactive'], 'Status');
      expect(() => enumChecker('Other', true, '')).toThrowError('is not a valid member of enumeration Status');
    });
  });

  describe('assertS3', () => {
    it('should produce a correct json formated string from an S3 value', () => {
      const s3 = {
        bucket: 'test',
        key: '58d33650-1f54-4190-b9f6-f85d0b889350',
        length: '1024',
        metadata: {
          type: 'personal document',
        },
      };
      expect(() => Assert.assertS3(s3)).not.toThrow();
      expect(Assert.assertS3(s3)).toEqual(s3);
      expect(Assert.assertS3(Assert.assertS3(s3))).toEqual(s3);
    });

    it('should throw if bad S3 value is provided', () => {
      const badS3 = {
        bucket: 'test',
        key: '58d33650-1f54-4190-b9f6-f85d0b889350',
        length: 'potato',
        metadata: {
          type: 'personal document',
        },
      };
      expect(() => Assert.assertS3(badS3 as any)).toThrowErrorMatchingSnapshot();
      expect(() => Assert.assertS3('{"key":"58d33650-1f54-4190-b9f6-f85d0b889350","length":1024,"metadata":{"type":"personal document"}}')).toThrowErrorMatchingSnapshot();
    });

    it('should throw if bad S3 value is provided which has missing properties', () => {
      const badS3 = {
        bucket: 'test',
        length: 'potato',
        metadata: {
          type: 'personal document',
        },
      };
      expect(() => Assert.assertS3(badS3 as any)).toThrowErrorMatchingSnapshot();
      expect(() => Assert.assertS3('{"length":1024,"metadata":{"type":"personal document"}}')).toThrowErrorMatchingSnapshot();
    });
  });
});
