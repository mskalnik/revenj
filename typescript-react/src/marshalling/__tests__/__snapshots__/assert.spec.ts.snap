// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`data marshalling asserting assertArray should explode on non-arrays 1`] = `"Expected an array, but value is null"`;

exports[`data marshalling asserting assertArray should explode on non-arrays 2`] = `"Expected an array, but value is [object Object]"`;

exports[`data marshalling asserting assertArray should explode on non-arrays 3`] = `"Expected an array, but value is [object Set]"`;

exports[`data marshalling asserting assertEnum should test that the value is one of the provided values 1`] = `"Value Other is not a valid member of enumeration Status"`;

exports[`data marshalling asserting assertEnum should test that the value is one of the provided values 2`] = `"Value INACTIVE is not a valid member of enumeration Status"`;

exports[`data marshalling asserting assertEnum should test that the value is one of the provided values 3`] = `"Value is required, but is not set!"`;

exports[`data marshalling asserting assertInt should explode on non-integers 1`] = `"Expected an Int, but value is -2147483649"`;

exports[`data marshalling asserting assertInt should explode on non-integers 2`] = `"Expected an Int, but value is bla"`;

exports[`data marshalling asserting assertInt should explode on non-integers 3`] = `"Expected an Int, but value is null"`;

exports[`data marshalling asserting assertInt should explode on non-integers 4`] = `"Expected an Int, but value is 1.55"`;

exports[`data marshalling asserting assertInt should explode on non-integers 5`] = `"Expected an Int, but value is 2147483648"`;

exports[`data marshalling asserting assertLong should assert valid Long range 1`] = `"Expected a Long, but value is -9223372036854775809"`;

exports[`data marshalling asserting assertLong should assert valid Long range 2`] = `"Expected a Long, but value is 9223372036854775808"`;

exports[`data marshalling asserting assertLong should explode on non-longs 1`] = `"Expected a Long, but value is bla"`;

exports[`data marshalling asserting assertLong should explode on non-longs 2`] = `"Expected a Long, but value is null"`;

exports[`data marshalling asserting assertLong should explode on non-longs 3`] = `"Expected a Long, but value is 1.55"`;

exports[`data marshalling asserting assertLong should explode on non-longs 4`] = `"Expected a Long, but value is 1.7976931348623157e+308"`;

exports[`data marshalling asserting assertLong should explode on non-longs 5`] = `"Expected a Long, but value is 5e-324"`;

exports[`data marshalling asserting assertLong should explode on non-longs 6`] = `"Expected a Long, but value is Infinity"`;

exports[`data marshalling asserting assertLong should explode on non-longs 7`] = `"Expected a Long, but value is -Infinity"`;

exports[`data marshalling asserting assertPresence should raise an error if the value is not present, but required 1`] = `"Value is required, but is not set!"`;

exports[`data marshalling asserting assertPresence should raise an error if the value is not present, but required 2`] = `"Value is required, but is not set!"`;

exports[`data marshalling asserting assertS3 should throw if bad S3 value is provided 1`] = `"S3 length: Expected a Long, but value is potato"`;

exports[`data marshalling asserting assertS3 should throw if bad S3 value is provided 2`] = `"S3 key: Expected a String, but value is undefined"`;

exports[`data marshalling asserting assertS3 should throw if bad S3 value is provided which has missing properties 1`] = `"S3 key: Expected a String, but value is undefined"`;

exports[`data marshalling asserting assertS3 should throw if bad S3 value is provided which has missing properties 2`] = `"S3 key: Expected a String, but value is undefined"`;

exports[`data marshalling asserting assertShort should assert valid Short range 1`] = `"Number -32769 out of Short range"`;

exports[`data marshalling asserting assertShort should assert valid Short range 2`] = `"Number 32768 out of Short range"`;

exports[`data marshalling asserting assertShort should explode on non-shorts 1`] = `"Expected an Int, but value is bla"`;

exports[`data marshalling asserting assertShort should explode on non-shorts 2`] = `"Expected an Int, but value is null"`;

exports[`data marshalling asserting assertShort should explode on non-shorts 3`] = `"Expected an Int, but value is 1.55"`;

exports[`data marshalling asserting assertString should also validate length if specified 1`] = `"Maximum length is 3, but received string of length 7"`;

exports[`data marshalling asserting assertString should also validate length if specified 2`] = `"Maximum length is 3, but received string of length 9"`;

exports[`data marshalling asserting assertString should explode on non-strings 1`] = `"Expected a String, but value is null"`;

exports[`data marshalling asserting assertString should explode on non-strings 2`] = `"Expected a String, but value is 12"`;

exports[`data marshalling asserting assertString should explode on non-strings 3`] = `"Expected a String, but value is [object Object]"`;

exports[`data marshalling asserting deserializeBinary should explode on non-strings 1`] = `"Expected a String, but value is null"`;

exports[`data marshalling asserting deserializeBinary should explode on non-strings 2`] = `"Expected a String, but value is 2"`;

exports[`data marshalling asserting deserializeBinary should explode on non-strings 3`] = `"Expected a String, but value is [object Object]"`;

exports[`data marshalling asserting deserializeMap should explode on non-lookups 1`] = `"Expected a dictionary object, but value is null"`;

exports[`data marshalling asserting deserializeMap should explode on non-lookups 2`] = `"Expected a dictionary object, but value is "`;

exports[`data marshalling asserting deserializeMap should explode on non-lookups 3`] = `"Expected a dictionary object, but value is [object Map]"`;

exports[`data marshalling asserting deserializeMap should explode on non-lookups 4`] = `"Expected a dictionary object, but value is [object Set]"`;

exports[`data marshalling asserting deserializeMap should explode on non-lookups 5`] = `
"Expected a dictionary object, but value is () => {
        return;
      }"
`;

exports[`data marshalling asserting deserializeResultSet should explode on invalid shapes 1`] = `"Expected an array, but value is [object Object]"`;

exports[`data marshalling asserting deserializeResultSet should explode on invalid shapes 2`] = `"Expected a result set of shape [columns[], types[], ...values[]], but received [object Map]"`;

exports[`data marshalling asserting deserializeResultSet should explode on invalid shapes 3`] = `"Expected a result set of shape [columns[], types[], ...values[]], but received "`;

exports[`data marshalling asserting deserializeSet should explode on non-arrays 1`] = `"Expected an array, but value is null"`;

exports[`data marshalling asserting deserializeSet should explode on non-arrays 2`] = `"Expected an array, but value is [object Object]"`;

exports[`data marshalling asserting deserializeSet should explode on non-arrays 3`] = `"Expected an array, but value is [object Set]"`;

exports[`data marshalling asserting serializeBinary should explode on non-strings 1`] = `"Expected a String, but value is null"`;

exports[`data marshalling asserting serializeBinary should explode on non-strings 2`] = `"Expected a String, but value is 2"`;

exports[`data marshalling asserting serializeBinary should explode on non-strings 3`] = `"Expected a String, but value is [object Object]"`;

exports[`data marshalling asserting serializeMap should explode on non-maps 1`] = `"Expected a Map, but value is null"`;

exports[`data marshalling asserting serializeMap should explode on non-maps 2`] = `"Expected a Map, but value is "`;

exports[`data marshalling asserting serializeResultSet should explode on invalid shapes 1`] = `"Expected a ResultSet, but received [object Object]"`;

exports[`data marshalling asserting serializeResultSet should explode on invalid shapes 2`] = `"Expected a ResultSet, but received [object Object]"`;

exports[`data marshalling asserting serializeResultSet should explode on invalid shapes 3`] = `"Expected a ResultSet, but received "`;

exports[`data marshalling asserting serializeSet should explode on non-sets 1`] = `"Expected a Set, but value is null"`;

exports[`data marshalling asserting serializeSet should explode on non-sets 2`] = `"Expected a Set, but value is [object Object]"`;

exports[`data marshalling asserting serializeSet should explode on non-sets 3`] = `"Expected a Set, but value is "`;
